### PR TITLE
Add unit-test for goto symex state assignments

### PR DESCRIPTION
**Pull request by romainbrenguier**
This adds unit test for state.assignment, and a couple of clean-up for things I noticed while looking at the function.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [na] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-symex/goto_symex_state.cpp [hunk 1]
```diff
@@ -193,8 +193,7 @@ renamedt<ssa_exprt, L2> goto_symex_statet::assignment(
     throw unsupported_operation_exceptiont(
       "pointer handling for concurrency is unsound");
 
-  // for value propagation -- the RHS is L2
-
+  // Update constant propagation map -- the RHS is L2
```
Comment (smowton) [inline src/goto-symex/goto_symex_state.cpp:196]: Belongs in different commit

File: unit/goto-symex/goto_symex_state.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+SCENARIO(
+  "Goto symex state assignment",
+  "[core][goto-symex][goto-symex-state][assignment]")
+{
+  // Initialize goto state
+  std::list<goto_programt::instructiont> target;
+  symex_targett::sourcet source{"fun", target.begin()};
+  guard_managert manager;
+  std::size_t fresh_name_count = 1;
```
Comment (smowton) [inline unit/goto-symex/goto_symex_state.cpp:39]: Static variable inside the lambda is usual

**Review by smowton — APPROVED** [review]

File: unit/goto-symex/goto_symex_state.cpp (hunk 2 above)
Comment (romainbrenguier) [reply unit/goto-symex/goto_symex_state.cpp:39]: Yes but when ran in catch that makes its result a bit unpredictable because the same counter will be used in each test but we don't have a guarantee on the order, so I wouldn't be able to use assertions like `result2.get().get_identifier() == "foo!0#2"`

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 81aec47172121ae142e1a124415e30cdbb98076e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116391013

**Review by forejtv — APPROVED** [review]
